pkg/telemetry/stargate_prometheus_resourcer.go

```diff
@@ -0,0 +1,252 @@
+        targetLabel: __name__
+      - regex: persistence_cassandra_(\d_\d+)_org_apache_cassandra_metrics_ClientRequest_ContentionHistogram_(Read|Write|CASRead|CASWrite|RangeSlice|ViewWrite)_count
+        replacement: stargate_client_request_contention_histogran_total
+        sourceLabels:
+        - __name__
+        targetLabel: __name__
+`
+
+var stargateEndpointHolder promapi.ServiceMonitor
+
+func init() {
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	_, _, err := decode([]byte(stargateEndpointString), nil, &stargateEndpointHolder)
+	if err != nil {
+		fmt.Println("Fatal error initialising EndpointHolder in pks/telemetry/stargate_prometheus_resourcer.go", err)
+		os.Exit(1)
+	}
+}
+
+// mustLabels() returns the set of labels essential to managing the Prometheus resources. These should not be overwritten by the user.
+func (cfg StargatePrometheusResourcer) mustLabels() map[string]string {
+	return map[string]string{
+		k8ssandraapi.ManagedByLabel:  k8ssandraapi.NameLabelValue,
+		k8ssandraapi.PartOfLabel:     k8ssandraapi.PartOfLabelValue,
+		k8ssandraapi.DatacenterLabel: cfg.StargateName,
+		stargateapi.StargateLabel:    cfg.StargateName,
+		k8ssandraapi.ComponentLabel:  k8ssandraapi.ComponentLabelTelemetry,
+		k8ssandraapi.CreatedByLabel:  k8ssandraapi.CreatedByLabelValueK8ssandraClusterController,
+	}
+}
+
+// NewServiceMonitor returns a Prometheus operator ServiceMonitor resource.
+func (cfg StargatePrometheusResourcer) NewServiceMonitor() (*promapi.ServiceMonitor, error) {
+	// validate the object we're being passed.
+	if cfg.StargateNamespace == "" || cfg.ServiceMonitorName == "" || cfg.StargateName == "" {
+		return nil, TelemetryConfigIncomplete{}
+	}
+	// Overwrite any CommonLabels the user has asked for if they conflict with the labels essential for the functioning of the operator.
+	mergedLabels := utils.MergeMap(cfg.CommonLabels, cfg.mustLabels())
+	stargateEndpointHolder.Spec.Endpoints[0].MetricRelabelConfigs[0].Replacement = cfg.StargateName
+	sm := promapi.ServiceMonitor{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ServiceMonitor",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cfg.ServiceMonitorName,
+			Namespace: cfg.StargateNamespace,
+			Labels:    mergedLabels,
+		},
+		Spec: promapi.ServiceMonitorSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					k8ssandraapi.ManagedByLabel: "k8ssandra-operator",
+					stargateapi.StargateLabel:   cfg.StargateName,
+				},
+			},
+			NamespaceSelector: promapi.NamespaceSelector{
+				MatchNames: []string{cfg.StargateNamespace},
+			},
+			Endpoints: stargateEndpointHolder.Spec.Endpoints,
+		},
+	}
+	annotations.AddHashAnnotation(&sm)
+	return &sm, nil
+}
+
+// UpdateResources executes the creation of the desired Prometheus resources on the cluster.
+func (cfg StargatePrometheusResourcer) UpdateResources(ctx context.Context, client runtimeclient.Client, owner *stargateapi.Stargate) error {
```

Comment:
This function is virtually the same as `CassPrometheusResourcer.UpdateResources`. With a bit of refactoring a lot of duplicate code can be eliminated.

Reply:
Done.